File: problem002/euler2.java
public class euler2{
	public static void main(String[] args){
		int x = 2;
		int total = 0;
		int i = fib(x);
		while (fib(x) < 4000000) {
			total += i;
			x += 3;
			i = fib(x);
		}
		System.out.println(total);
	}
	private static int[] a;
	private static int fib(int n){
		if (a == null) {
			if (n < 2) {
				return 1;
			} else {
				return fib(n - 1) + fib(n - 2);
			}
		} else {
			return a[n];
		}
	}
}

File: problem002/euler2.scala
object euler2 {
    def fib_iter(a: Int, b: Int, count: Int) : Int = {
        count match {
            case 0 => b
            case _ => fib_iter(a + b, a, count - 1)
        }
    }
    
    def fibonacci(n: Int): Int = {
        return fib_iter(1, 0, n);
    }

    def sumEvenValuedTermsUnder(start: Int, limit: Int, acc: Int) : Int = {
        val n = fibonacci(start)
        if (n >= limit)
            acc
        else
            sumEvenValuedTermsUnder(start + 3, limit, acc + n)
    }

    def main(args: Array[String]) {
        println(sumEvenValuedTermsUnder(3, 4000000, 0));
    }
}
